Examples.md

```diff
@@ -29,7 +31,20 @@ async with grpc.insecure_channel('localhost:50051') as channel:
     response = await stub.Hi(...)
 ```
 
-### Ideal
+### Option 3
+
+```Python
+ctx = grpc.Context() \
+    .with_timeout_secs(5.0) \
+    .append_outgoing_metadata('key', 'value')
```

Comment:
Any reason for this builder pattern style over kwargs? It seems they'd accomplish the same thing.

Reply:
`grpc.Context` is supposed to be passed through method boundaries, each method implementation can augment it by creating new context, example:

```
class BarServicer(object):
  async def foo1(self, ctx: grpc.Context) -> foo_pb.Response:
    return await self.client.Foo(ctx.with_timeout_secs(5.0), foo_pb.Request())

  async def foo2(self, ctx: grpc.Context) -> foo_pb.Response:
    if 'request_id' not in ctx.inbound_metadata():
      ctx = ctx.append_outgoing_metadata('request_id', request_id.random_new())
    return await foo1(ctx)

  async def BarHandler(self, ctx: grpc.Context, req: bar_pb.Request) -> bar_pb.Response:
    return self.construct_bar_response_from_foo(await self.foo1(ctx))
```

It's designed after [golang Context](https://golang.org/pkg/context/) and provides roughly similar functionality.

Reply:
Have you taken a look at https://docs.python.org/3/library/contextvars.html? Any thoughts on that approach versus this one?

Reply:
This API didn't exist 2-3 years ago when were adding a notion of `Context` internally.

Looking at it today it can be viable, though I don't see any protections against mutating the same `contextvars.Context` object from concurrent asyncio co-routines within the same server call, which makes it dangerous/hard to troubleshoot.

For example:
1. Server receives new rpc
2. Handler calls 2 concurrent async methods and awaits for them using `asyncio.gather`
3. Each of this concurrent calls somehow alters the context (e.g. to tighten deadline, add some outgoing metadata, etc) but they alter the same context so state leaks across co-routines
4. Some engineer spends hours trying to troubleshoot why deadline is different or some metadata is wrong

With immutable append-only `Context` this situation is impossible as each co-routine will pass along its own Context object and its alteration will just create a new object and keep parent unmodified.

Reply:
I believe `contextvars` handle the situation you described well. They are copy-on-write objects, with a concept of parents that can be and are restored by the `asyncio` event loop. Consider the following snippet:

```python
import asyncio
import contextvars
import random


cv = contextvars.ContextVar('grpc.deadline')
cv.set(0)

async def task1():
    await asyncio.sleep(random.random())
    print(f"Task 1 deadline before: {cv.get()}")
    cv.set(1)
    await asyncio.sleep(random.random())
    print(f"Task 1 deadline after: {cv.get()}")

async def task2():
    await asyncio.sleep(random.random())
    print(f"Task 2 deadline before: {cv.get()}")
    cv.set(2)
    await asyncio.sleep(random.random())
    print(f"Task 2 deadline after: {cv.get()}")

async def task3():
    await asyncio.sleep(random.random())
    print(f"Task 3 deadline before: {cv.get()}")
    cv.set(3)
    await asyncio.sleep(random.random())
    print(f"Task 3 deadline after: {cv.get()}")

async def main():
    await asyncio.gather(
        task1(),
        task2(),
        task3(),
    )
    print(f"Final deadline: {cv.get()}")

asyncio.run(main())
```

The result is what you would hope:
```
Task 3 deadline before: 0
Task 3 deadline after: 3
Task 1 deadline before: 0
Task 2 deadline before: 0
Task 1 deadline after: 1
Task 2 deadline after: 2
Final deadline: 0
```

Reply:
This actually looks promising.

Reply:
Having the feeling that we are replacing entirely the traditional arguments for a whole context object. What's the rationale for doing so? Moving across gRPC services information/values that are also used by the gRPC library?

Also regarding the context by itself, leaving aside my first skeptical opinion on using it, I have two comments:

- Requiring it explicitly as a parameter is IMO good idea, that would prevent having future bugs because the user didn't remember that the context was implicitly used.
- Regarding the usage of the `contextvars` it would make sense IMO for giving access to request context to any piece of the code. But the user would be still forced to explicitly give the parameter when is required by a gRPC function.

Having the feeling that where you are saying `context` IMO it means `request`,  where none gRPC data can be placed within an attribute of the `request`, for example `request.user_data`.

So this would bring us to have something like:

```python
class BarServicer(object):
  async def foo1(self, request: grpc.Request) -> foo_pb.Response:
    foo_request = grpc.Request.from_request(request)
    return await self.client.Foo(foo_request.with_timeout_secs(5.0), foo_pb.Request())
```

Which IMO is another way of implementing the current pattern that uses arguments.

Reply:
gRPC has not concept of `request`-params, but it has a concept of `context`.
Please check gRPC API in other languages (golang, java, c++, etc), this object is always named `context`.

> Having the feeling that we are replacing entirely the traditional arguments for a whole context object. What's the rationale for doing so? Moving across gRPC services information/values that are also used by the gRPC library?

There are few reasons:
1. It matches semantics of gRPC implementation in all other languages
2. It aligns better with the nature/usages of this properties, e.g. `timeout` itself has little value on per-method basis, usually services are written with `deadline` in mind, where `deadline` is a property of the parent context and engineers rely on its automatic propagation across the stack. Similar logic is applicable to tracing, outbound metadata, etc.